psyneulink/core/components/functions/userdefinedfunction.py

```diff
@@ -29,6 +30,295 @@
 
 __all__ = ['UserDefinedFunction']
 
+class UserDefinedFunctionVisitor(ast.NodeVisitor):
+    def __init__(self, udf, ctx, builder, params, state, arg_in, arg_out, *, tags:frozenset):
+        self.udf = udf
+        self.ctx = ctx
+        self.builder = builder
+        self.params = params
+        self.state = state
+        self.arg_in = arg_in
+        self.arg_out = arg_out
+        self.tags = tags
+        self.register = {}
+        #setup default functions
+        from psyneulink.library.compositions.pytorchllvmhelper import gen_inject_vec_sum
+        self.register['sum'] = lambda x: gen_inject_vec_sum(ctx, builder, x)
+
+        self._register_numpy()
+        super().__init__()
+
+    def _register_numpy(self):
+        numpy_handlers = {
+            'tanh': lambda x: pnlvm.helpers.call_elementwise_operation(self.ctx, self.builder, x, pnlvm.helpers.tanh),
+            'exp': lambda x: pnlvm.helpers.call_elementwise_operation(self.ctx, self.builder, x, pnlvm.helpers.exp)
+        }
+        func_globals =self.udf.custom_function.__globals__
+        for k,v in func_globals.items():
+            if v is np:
+                self.register[k] = numpy_handlers
+
+    def visit_arguments(self, node):
+        args = node.args
+        variable = args[0]
+        # update register
+        self.register[variable.arg] = self.arg_in
+        parameters = args[1:]
+        for param in parameters:
+            assert param.arg not in ["self", "owner"], f"Unable to reference {param.arg} in a compiled UserDefinedFunction!"
+            if param.arg == 'params':
+                self.register[param.arg] = self.params
+            elif param.arg == 'context':
+                # HACK: Since contexts are implicit in the structs in compiled mode, we do not compile it.
+                pass
+            else:
+                self.register[param.arg] = pnlvm.helpers.get_param_ptr(self.builder, self.udf, self.params, param.arg)
+
+    def _get_type(self, t):
+        type_t = getattr(t, "type", t)
+        if type(type_t) is pnlvm.ir.DoubleType:
+            return "scalar"
+        elif type_t is pnlvm.ir.IntType(1):
```

Comment:
can't use isinstance for bool checking, since pnlvm.ir.IntType(1) isn't a valid python type; the given syntax seems to be the only way to do this check.

Reply:
does `isinstance(type_t, IntType) and type_t.width == 1` not work?

Reply:
it does - fixed